### ISSUE TITLE
Prompt not updating with SIMBA

### ISSUE THREAD

**Issue by sky-2002** [stanfordnlp/dspy#8592]
I am trying SIMBA optimizer on a very small dataset of 8 examples.

```python
class Summarization(dspy.Signature):
    """Summarize a ist of question-answer pairs into a neatly formatted and readable
    section without missing information"""

    qna: str = dspy.InputField(
        desc="""A concatenated list of question-answer pairs"""
    )
    
    summary: str = dspy.OutputField(
        desc="""A neatly formatted and readable section summarizing the information present in question-answer pairs"""
    )
```

Attaching a sample evaluation code:
```python
from pydantic import BaseModel, Field
import openai

openai_client = openai.Client()

class Eval(BaseModel):
    operational_performance: bool
    production_volumes: bool
    etc

    def coverage_ratio(self):
        total_items = len(self.__dict__)
        covered_items = sum(1 for value in self.__dict__.values() if value)
        return covered_items / total_items if total_items > 0 else 0.0


eval_prompt = """
Given the "performance highlights of the year section content,
evaluate it based on whether the following checklist items are covered in the section or not.

**Checklist for "<SOME SECTION>":**

<SOME CHECKLIST>

**In the output, provide whether each item is covered or not. True if covered, False if not covered.**

**Section content:**
{section_content}
"""


def score_summary(example, predicted_summary):
    score = None
    feedback = None

    _eval_predicted = openai_client.responses.parse(
        model="gpt-4.1-mini",
        input=[
            {"role": "system", "content": "You are an expert in evaluating business reports."},
            {"role": "user", "content": eval_prompt.format(section_content=predicted_summary)},
        ],
        text_format=Eval
    ).output_parsed

    _score = _eval_predicted.coverage_ratio()
    

    return dspy.Prediction(
        score=_score,
    )
```

Program and training:
```python
dspy_examples = []
for i in range(len(data)):
    s = data.iloc[i]['summary']
    qna = data.iloc[i]['qna']
    dspy_examples.append(
        dspy.Example(
            qna=qna,
        ).with_inputs("qna")
    )
program = dspy.ChainOfThought(Summarization)
optimizer = dspy.SIMBA(
    metric=score_summary,
    bsize=4,
    num_candidates=2,
    max_steps=1
)
summarizer_with_simba = optimizer.compile(
    program,
    trainset= dspy_examples,
)
```
Below is what I get at the end of compilation:
<img width="1626" height="87" alt="Image" src="https://github.com/user-attachments/assets/1a71e0ed-54fd-4866-9048-93663ff4d0bb" />

I also see that the prompts are not updated after training.
<img width="1548" height="134" alt="Image" src="https://github.com/user-attachments/assets/f854b8dc-0ec3-48c1-983b-be2f22d76527" />

Also "train" is empty when i save optimized program:

<img width="1301" height="182" alt="Image" src="https://github.com/user-attachments/assets/d44cc228-cc32-480b-bb3d-ce189fea2c00" />

Am I doing something wrong? or missing something? or are there very few examples? There could be issue with my metric too, but some pointers or help would be great.

@krypticmouse @okhat

**Comment by TomeHirata** [comment]
I think the core reason for the result is the very small size of the training set. According to the log, it seems only two candidate programs are tried and the first candidate (the initial program) shows the highest performance, resulting in returning the initial program as the final program. Can you try increasing the train set size? The default batch size of SIMBA is 32.

**Comment by a-alzboon** [comment]
Same here. the instructions were being optimized: rules and  demos were logged during the optimization process, but the optimizer json state are the same as default without demos or improved instructions.
